pkg/fqdn/service/service_test.go

```diff
@@ -0,0 +1,261 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/netip"
+	"testing"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/hivetest"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/cilium/cilium/pkg/clustermesh/types"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/endpoint"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/fqdn/messagehandler"
+	"github.com/cilium/cilium/pkg/fqdn/namemanager"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/ipcache"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/policy/api"
+	"github.com/cilium/cilium/pkg/testutils"
+	testidentity "github.com/cilium/cilium/pkg/testutils/identity"
+	"github.com/cilium/cilium/pkg/time"
+
+	pb "github.com/cilium/cilium/api/v1/standalone-dns-proxy"
+)
+
+func TestFQDNDataServer(t *testing.T) {
+
+	test := map[string]struct {
+		port                     int
+		enableL7Proxy            bool
+		enableStandaloneDNSProxy bool
+		serverPort               int
+		err                      error
+	}{
+		"Successfully running the server": {
+			port:                     1234,
+			enableL7Proxy:            true,
+			enableStandaloneDNSProxy: true,
+			serverPort:               40045,
+			err:                      nil,
+		},
+		"Failure on running the server due to invalid server port": {
+			port:                     1234,
```

Comment:
Choosing a specific port to listen on is going to be flaky: whenever some other test has randomly chosen this port to listen, this test will fail to bind. The chance is low, but at the scale of cilium CI it'll happen often enough to be annoying.

I believe the proper way to fix this is to use an in-memory transport. Luckily gRPC provides this: https://pkg.go.dev/google.golang.org/grpc/test/bufconn.

Reply:
I see. The test is supposed to see if the server can listen on the given port. Shouldn't this be used when we add the other method(sending data to SDP and receiving data from sdp) ?
Currently, the test is for checking if we are able to start the server.(testing job oneShot). Once we add the other method, I can use the bufcon.(I already used that in the parent PR: https://github.com/cilium/cilium/pull/36213/files#diff-63c75a66eba03ca699354a9d93ebad39fc014606857a52db57c742dd01e8d810R21)

Reply:
We can't merge a PR knowing that it makes CI more flaky. If the bufcon isn't what you want to use at this point, that's fine, but you'll need to figure out a way which doesn't rely on choosing a specific port that could already be taken.

Maybe the way to go is finding a free port (by listening on 0) and then using that, i.e. fixing up the config after. I think that requires about the same amount of work than using a bufcon, though.

Reply:
Understood. I updated the test scenario. Removed the zero port test case, as we are now using `bufcon` and no way to check the logs.(as we are adding the log if port is zero)